### PR TITLE
Data Wrangler (feature PR)

### PR DESCRIPTION
**Pull request by vandyliu**
This is a feature branch that I will be merging PRs into, then in the end merging it to main.

In the original PR, there are the data wrangler parts, slick grid, webview, the command palette commands. 
I have removed the side panels so it is a little easier to review.

I put `data-wrangler` under `data-viewer` because data-wrangler uses quite a few stuff from `data-viewer` but I'm open to comments/opinions.

PR review video availabe in Data Wrangler Feature Branch PR Call teams chat

### CODE REVIEW

File: src/datascience-ui/data-explorer/data-wrangler/reactSlickGrid.tsx [hunk 1]
```diff
@@ -0,0 +1,802 @@
...
+                    {/* <Resizable 
+                        style={{display: "flex", alignItems: "top", justifyContent: "left", flexDirection: "column", zIndex: 99998 }}
+                        handleClasses={{ right: "resizable-span" }}
+                        defaultSize={{ width: '60%', height }}
+                        enable={{ left:false, top:false, right:true, bottom:false, topRight:false, bottomRight:false, bottomLeft:false, topLeft:false }}
+                        > */}
+                    <div className="react-grid-container" style={style} ref={this.containerRef}></div>
+                    <div className="react-grid-measure" ref={this.measureRef} />
+                    <Resizable>
+                        {/* <Resizable
```
Comment (vandyliu) [inline src/datascience-ui/data-explorer/data-wrangler/reactSlickGrid.tsx:339]: Side Panel (will be in another PR)

File: src/client/common/application/customEditorService.ts [hunk 2]
```diff
@@ -36,10 +36,11 @@ export class CustomEditorService implements ICustomEditorService, IExtensionSing
         if (this.useCustomEditorApi) {
             // eslint-disable-next-line @typescript-eslint/no-explicit-any
             return (vscode.window as any).registerCustomEditorProvider(viewType, provider, options);
-        } else {
+        } else if (!this.useCustomEditorApi) {
```
Comment (rchiodo) [inline src/client/common/application/customEditorService.ts:39]: Weird? How can a boolean be both not true and not false?

File: src/client/datascience/commands/commandRegistry.ts [hunk 3]
```diff
@@ -552,3 +562,26 @@ export class CommandRegistry implements IDisposable {
         }
     }
 }
+
+export function getImportCodeForFileType(filepath: string) {
```
Comment (rchiodo) [inline src/client/datascience/commands/commandRegistry.ts:566]: Seems like a weird spot for this? Not sure what it has to do with the commandRegistry?

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+    IDropNaRequest,
+    OpenDataWranglerSetting,
+    IDataWranglerJupyterVariableDataProvider,
+    IDataWranglerJupyterVariableDataProviderFactory
+} from './types';
+
+const PREFERRED_VIEWGROUP = 'JupyterDataWranglerPreferredViewColumn';
+const dataWranglerDir = path.join(EXTENSION_ROOT_DIR, 'out', 'datascience-ui', 'viewers');
+@injectable()
+export class DataWrangler extends WebviewPanelHost<IDataWranglerMapping> implements IDataWrangler, IDisposable {
```
Comment (rchiodo) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:72]: I'm wondering if this class and the DataViewer can share anything? They are pretty similar.

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 5]
```diff
@@ -0,0 +1,685 @@
...
+            }
+        }
+    }
+
+    private renameColumn(currentVariableName: string, req: IRenameColumnsRequest): IHistoryItem {
+        this.variableCounter += 1;
+        const newVariableName = `df${this.variableCounter}`;
+        const code = `${newVariableName} = ${currentVariableName}.rename(columns={ "${req.oldColumnName}": "${req.newColumnName}" })\n`;
+        const historyItem = {
+            transformation: `Renamed column "${req.oldColumnName}" to "${req.newColumnName}"`,
```
Comment (rchiodo) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:565]: I believe these 'actions' need to be localized.

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 6]
```diff
@@ -0,0 +1,685 @@
...
+            case DataWranglerCommands.GetHistoryItem:
+                this.getHistoryItem(payload.args.index).ignoreErrors();
+                break;
+
+            case DataWranglerCommands.Describe:
+                void this.getColumnStats(payload.args.columnName);
+                break;
+        }
+
+        const dataCleaningMode = this.configService.getSettings().dataCleaningMode;
```
Comment (rchiodo) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:520]: You might abstract out this two pronged approach into separate classes that implement an interface? DataWranglerCommandHandler or something like that? One that executes in a notebook. One that is standalone?

Comment (rchiodo) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:520]: Or just two functions at least with the same signature.

Comment (rchiodo) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:520]: I think the commands could also be a map from command to function (instead of the big switch statement), each function returning a historyItem or undefined if no further history update required.

File: src/client/datascience/data-viewing/data-wrangler/dataWranglerJupyterVariableDataProvider.ts [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+import { Identifiers } from '../../constants';
+import { IJupyterVariable, IJupyterVariables, INotebook } from '../../types';
+import { DataViewerDependencyService } from '../dataViewerDependencyService';
+import { ColumnType, IDataFrameInfo, IRowsResponse, IColsResponse } from '../types';
+import { traceError } from '../../../common/logger';
+import { IDataWranglerJupyterVariableDataProvider } from './types';
+
+@injectable()
+export class DataWranglerJupyterVariableDataProvider implements IDataWranglerJupyterVariableDataProvider {
```
Comment (rchiodo) [inline src/client/datascience/data-viewing/data-wrangler/dataWranglerJupyterVariableDataProvider.ts:16]: Is this much different than IJupyterVariables? Why not add the new methods to IJupyterVariables and just use those? 

That would have the added benefit of supporting doing this during debugging too (although that's not a likely use case)

File: src/client/datascience/data-viewing/data-wrangler/dataWranglerProvider.ts [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+        if (dataCleaningMode == '') {
+            const qpoptions: QuickPickOptions = {
+                ignoreFocusOut: false,
+                matchOnDescription: true,
+                matchOnDetail: true
+            };
+
+            const qpitems = [
+                {
+                    label: 'Open Just the Data Wrangler',
```
Comment (rchiodo) [inline src/client/datascience/data-viewing/data-wrangler/dataWranglerProvider.ts:120]: Localization for these strings.

File: src/datascience-ui/data-explorer/data-wrangler/mainPanel.tsx [hunk 9]
```diff
@@ -0,0 +1,632 @@
...
+    //                     ) : undefined}
+    //                     <span>{breadcrumbText}</span>
+    //                 </div>
+    //             </div>
+    //         );
+    //     }
+    // }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    public handleMessage = (msg: string, payload?: any) => {
```
Comment (rchiodo) [inline src/datascience-ui/data-explorer/data-wrangler/mainPanel.tsx:245]: I wonder if we can abstract out part of this. A lot of the webviews have to handle the onigasm stuff. Seems like we don't need to repeat it in each.

File: src/datascience-ui/data-explorer/data-wrangler/reactSlickGrid.tsx [hunk 10]
```diff
@@ -0,0 +1,802 @@
...
+    handleRefreshRequest(): void;
+}
+
+interface ISlickGridState {
+    grid?: Slick.Grid<ISlickRow>;
+    showingFilters?: boolean;
+    fontSize: number;
+}
+
+export class ReactSlickGrid extends React.Component<ISlickGridProps, ISlickGridState> {
```
Comment (rchiodo) [inline src/datascience-ui/data-explorer/data-wrangler/reactSlickGrid.tsx:97]: Yeah this has to be reusable. It looks identical to the one in the data viewer.

**Review by rchiodo — CHANGES_REQUESTED** [review]
Looks pretty good. I think there are some spots for code reuse though. Open to discussion on them of course. Or some other way to achieve reuse.

File: src/client/datascience/commands/commandRegistry.ts [hunk 11]
```diff
@@ -477,14 +482,18 @@ export class CommandRegistry implements IDisposable {
         }
     }
 
-    private async createNewNotebook(): Promise<void> {
+    private async createNewNotebook(): Promise<INotebookEditor | undefined> {
         if (this.useNativeNotebook) {
-            await this.nativeNotebookCreator.createNewNotebook();
+            return await this.nativeNotebookCreator.createNewNotebook();
```
Comment (IanMatthewHuff) [inline src/client/datascience/commands/commandRegistry.ts:487]: Isn't return await redundant? Thought there was an eslint rule for that.

File: package.json [hunk 12]
```diff
@@ -1825,6 +1873,16 @@
                     }
                 ],
                 "priority": "option"
+            },
+            {
+                "viewType": "jupyter-data-wrangler",
+                "displayName": "Data Wrangler",
+                "selector": [
+                    {
+                        "filenamePattern": "*.csv"
```
Comment (IanMatthewHuff) [inline package.json:1882]: Are the other file types needed here? Not sure until I read further in the review.

File: src/client/datascience/commands/commandRegistry.ts [hunk 13]
```diff
@@ -552,3 +562,26 @@ export class CommandRegistry implements IDisposable {
...
 }
+
+export function getImportCodeForFileType(filepath: string) {
+    const fileExtension = path.extname(filepath);
+    let code = 'import pandas as pd\n';
+    switch (fileExtension) {
+        case '.csv':
+            code += `df = pd.read_csv(r"${filepath}")`;
+            break;
+        case '.xlsx': // TODO dependency check for openpyxl
```
Comment (IanMatthewHuff) [inline src/client/datascience/commands/commandRegistry.ts:573]: TODOs are ok, but are they being tracked somewhere? They always make me nervous going in, easy to forget. If you have more might want to tag them for easy searching. Like (DataWranglerTODO).

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts (hunk 5 above)
Comment (vandyliu) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts]: Added to localize.ts and package.nls.json

File: src/client/datascience/commands/commandRegistry.ts (hunk 13 above)
Comment (vandyliu) [reply src/client/datascience/commands/commandRegistry.ts]: Good point, removed them

File: package.json (hunk 12 above)
Comment (vandyliu) [reply package.json:1882]: Will just be implementing csv for now

File: src/client/datascience/commands/commandRegistry.ts (hunk 11 above)
Comment (vandyliu) [reply src/client/datascience/commands/commandRegistry.ts:486]: Yep, it is and removed all instances of it. Ty

File: src/client/datascience/data-viewing/data-wrangler/dataWranglerJupyterVariableDataProvider.ts (hunk 7 above)
Comment (vandyliu) [reply src/client/datascience/data-viewing/data-wrangler/dataWranglerJupyterVariableDataProvider.ts]: I guess I was scared of adding more methods to the interface in case other components relied on them and would have to update them, but I see your point and will just add it there.

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts (hunk 4 above)
Comment (vandyliu) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:72]: They are definitely very similar. So are some of the other classes, I really really want to figure out the best way to re-use code.  Here are my thoughts: abstract classes, but it won't work because the classes `extend` the same class but the `Mapping` is different. Same for the React classes with the props. Unless I made it so they used the same mappings, and made the data wrangler ones optional. I could also have a dataViewingHelper (probably needs a better name) class that both dataViewer and dataWrangler call when the code is the same but there will still probably be a bit of duplicate code. Any other ideas or thoughts on the above? They are very welcomed and thank you for your review. I appreciate it a lot.

Comment (rchiodo) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:72]: DataWrangler could probably be a subclass of DataViewer on the extension side? Its handleMessage might have some extra message handlers but otherwise they seem identical. 

On the react side, the message handling could be abstracted into a base class for both (and others) or you could use some object to handle the messages.  Actually I think the tmlanguage and onigasm stuff is already abstracted using redux, but maybe the datawrangler/dataviewer don't bother with redux. If they don't, then just creating a common base class or some other object to handle those messages seems like it would work.

Comment (vandyliu) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts]: Alright, so I did the `extends` implementation
I decided to just add DataWranglerMessages to IDataViewerMapping
I played around with the constructor because I wasn't sure about the best way to get the different values into WebviewPanelHost<IDataViewerMapping> (Eg. title, directory strings). What do you think?

Comment (rchiodo) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts]: Couldn't DataWrangler extend DataViewer? They're very similar

Comment (vandyliu) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts]: Yep, that's what I did in my most recent commits.

File: src/client/datascience/jupyter/kernelVariables.ts [hunk 14]
```diff
@@ -152,6 +152,20 @@ export class KernelVariables implements IJupyterVariables {
         return {};
     }
 
+    public async getDataFrameColumn(
```
Comment (rchiodo) [inline src/client/datascience/jupyter/kernelVariables.ts:155]: 👍

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts (hunk 6 above)
Comment (vandyliu) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:375]: I see what you mean with the separate classes implementing an interface. It would especially be smart once other forms are added like WITH_INTERACTIVE_WINDOW, etc. It started getting tricky due to the async/awaits and disposables. For the two functions, I believe that doesn't work in typescript because of it getting compiled to javascript (or I may be misinterpreting what you mean). I do like the command to function idea, but there are some async/await functions which makes it a little awkward. I can make them all async. Do you think that would be worthwhile exploring?

Comment (rchiodo) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:375]: Making them all async sounds fine to me. Non async functions that are marked that way don't infer a penalty AFAIK. They just run synchronously.

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 15]
```diff
@@ -0,0 +1,502 @@
...
+
+            default:
+                break;
+        }
+
+        // Some messages will be handled by DataViewer
+        super.onMessage(message, payload);
+    }
+
+    private async requestTmLanguage(languageId: string = PYTHON_LANGUAGE) {
```
Comment (rchiodo) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:226]: We must be doing this somewhere else? Ah yes we are. InteractiveBase. I think you can move this up one level to WebviewPanelHost and then that would work for this code too.

Then all of the monaco init stuff will be in one spot.

Comment (rchiodo) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:226]: Meaning change InteractiveBase and this code to not have these monaco related functions, but instead move them to WebviewPanelHost.

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 16]
```diff
@@ -0,0 +1,502 @@
...
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private async handleCommand(payload: { command: string; args: any }) {
+        const notebook = (this.dataProvider as IJupyterVariableDataProvider).notebook;
+        let code = '';
+        const currentVariableName = (await this.dataFrameInfoPromise)!.name ?? '';
+        let newVariableName = currentVariableName ?? '';
+
+        // Get and run data wrangler command
+        const cmd = this.commands.get(payload.command as DataWranglerCommands);
```
Comment (rchiodo) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:316]: Much nicer 👍

File: src/client/datascience/data-viewing/data-wrangler/dataWranglerProvider.ts [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+import { Identifiers } from '../../constants';
+import { INotebookProvider, IJupyterVariables, IJupyterVariableDataProviderFactory } from '../../types';
+import { DataViewerChecker } from '../../interactive-common/dataViewerChecker';
+import { IConfigurationService } from '../../../common/types';
+import { IDataWranglerFactory } from './types';
+import { DataScience } from '../../../common/utils/localize';
+import { IDataViewerDataProvider } from '../types';
+
+@injectable()
+export class DataWranglerProvider implements IDataWranglerProvider, IExtensionSingleActivationService {
```
Comment (rchiodo) [inline src/client/datascience/data-viewing/data-wrangler/dataWranglerProvider.ts:25]: can you put a comment on the top of each class explaining what it's for? For example it's not clear immediately the difference between DataWranglerProvider and DataWranglerFactory

File: src/datascience-ui/data-explorer/data-wrangler/controls/toolbar.tsx [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+                        className="codicon codicon-go-to-file codicon-button"
+                        style={{ verticalAlign: 'middle' }}
+                        title="Open as Python script"
+                    />
+                    <span style={{ verticalAlign: 'middle', paddingLeft: '4px', paddingBottom: '4px' }}>
+                        Open as Python script
+                    </span>
+                    {/* <div className="codicon codicon-notebook codicon-button" title="Open in Notebook" /> */}
+                </div>
+                {/* <div
```
Comment (rchiodo) [inline src/datascience-ui/data-explorer/data-wrangler/controls/toolbar.tsx:57]: Why is a bunch of this commented out? Do we intend to enable this in the future?

File: src/datascience-ui/data-explorer/data-wrangler/mainPanel.tsx (hunk 9 above)
Comment (rchiodo) [reply src/datascience-ui/data-explorer/data-wrangler/mainPanel.tsx:249]: The state transitions in the datawrangler may warrant using redux instead of handling everything in the react code, but I think we can leave that till later/next pr.

File: src/test/datascience/data-viewing/dataWrangler.unit.test.ts [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+        await dataWrangler.showData(instance(dataProvider), title);
+
+        verify(dataProvider.getDataFrameInfo(anything(), anything())).once();
+    });
+    test('Data Wrangler calls data provider dispose', async () => {
+        await dataWrangler.showData(instance(dataProvider), title);
+        dataWrangler.dispose();
+
+        verify(dataProvider.dispose()).once();
+    });
```
Comment (rchiodo) [inline src/test/datascience/data-viewing/dataWrangler.unit.test.ts:91]: I think we should have an integration test for it? Maybe I missed that.

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts (hunk 15 above)
Comment (vandyliu) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:226]: The IMapping part is a little tricky, I'll message you about it.

File: src/test/datascience/data-viewing/dataWrangler.unit.test.ts (hunk 19 above)
Comment (vandyliu) [reply src/test/datascience/data-viewing/dataWrangler.unit.test.ts:91]: added a simple one

File: src/datascience-ui/data-explorer/data-wrangler/controls/toolbar.tsx (hunk 18 above)
Comment (vandyliu) [reply src/datascience-ui/data-explorer/data-wrangler/controls/toolbar.tsx]: still wasn't sure if they were going to be included or not but got confirmation and removed them now

File: package.json [hunk 20]
```diff
@@ -2044,20 +2108,25 @@
         "node-stream-zip": "^1.6.0",
         "onigasm": "^2.2.2",
         "pdfkit": "^0.12.1",
+        "plotly.js": "^1.58.4",
```
Comment (IanMatthewHuff) [inline package.json:2111]: Is this including plotly twice? This might just be one minified and one-non minified version.

Comment (vandyliu) [reply package.json]: yep, and fixed ty

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 21]
```diff
@@ -0,0 +1,461 @@
...
+            case DataWranglerMessages.SubmitCommand:
+                this.handleCommand(payload).ignoreErrors();
+                break;
+
+            case DataWranglerMessages.RefreshDataWrangler:
+                this.refreshData().ignoreErrors();
+                break;
+
+            case InteractiveWindowMessages.LoadTmLanguageRequest:
+                void this.requestTmLanguage(payload);
```
Comment (rchiodo) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:205]: Shouldn't these be handled by the super.handleMessage?

File: src/client/datascience/data-viewing/data-wrangler/dataWranglerFactory.ts [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+        const editor = window.activeTextEditor;
+        if (!editor) return;
+        const document = editor.document;
+        if (!document) return;
+        const position = editor.selection;
+        if (!position) return;
+        // See if a variable exists
+        // Look for an active data wrangler
+        if (this.knownDataWranglers.size === 0) {
+            // Create a new data wrangler
```
Comment (rchiodo) [inline src/client/datascience/data-viewing/data-wrangler/dataWranglerFactory.ts:118]: Seems like this case isn't actually being handled?

File: src/client/datascience/data-viewing/data-wrangler/dataWranglerProvider.ts (hunk 17 above)
Comment (rchiodo) [reply src/client/datascience/data-viewing/data-wrangler/dataWranglerProvider.ts:28]: Can you put this same comment on all of the new classes you added?

File: src/client/datascience/data-viewing/jupyterVariableDataProvider.ts [hunk 23]
```diff
@@ -47,10 +49,14 @@ export class JupyterVariableDataProvider implements IJupyterVariableDataProvider
...
                     normalizedType = ColumnType.String;
             }
-            return {
+            const col = {
                 key: column.key,
                 type: normalizedType
-            };
+            } as { key: string; type: ColumnType; describe?: string };
+            if (column.describe) {
+                col['describe'] = column.describe;
```
Comment (rchiodo) [inline src/client/datascience/data-viewing/jupyterVariableDataProvider.ts:57]: What happens if this column isn't here? Do you need a default value for this?

File: src/client/datascience/interactive-common/interactiveBase.ts [hunk 24]
```diff
@@ -329,6 +331,10 @@ export abstract class InteractiveBase extends WebviewPanelHost<IInteractiveWindo
                 this.handleMessage(message, payload, this.showDataViewer);
                 break;
 
+            case InteractiveWindowMessages.ShowDataWrangler:
+                this.handleMessage(message, payload, this.showDataWrangler);
+                break;
+
             case InteractiveWindowMessages.GetVariablesRequest:
                 this.handleMessage(message, payload, this.requestVariables);
                 break;
```
Comment (rchiodo) [inline src/client/datascience/interactive-common/interactiveBase.ts:340]: I was thinking the requestTmLanguage here would be in the base class. So you don't need to handle the message here.

Comment (rchiodo) [reply src/client/datascience/interactive-common/interactiveBase.ts:340]: the ```super.onMessage``` would handle it instead.

**Review by rchiodo — APPROVED** [review]
Just some minor comments about adding comments on all new classes and maybe not needing to handle the monaco messages in the sub classes.

But looks good to me. Feel free to submit.

File: src/client/datascience/data-viewing/jupyterVariableDataProvider.ts (hunk 23 above)
Comment (vandyliu) [reply src/client/datascience/data-viewing/jupyterVariableDataProvider.ts:57]: Not exactly sure what you meant. Do you mean if describe doesn't exist? If that's the case, it can be undefined, I could also make it the empty string

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts (hunk 21 above)
Comment (vandyliu) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:211]: yep

File: src/client/datascience/data-viewing/jupyterVariableDataProvider.ts (hunk 23 above)
Comment (rchiodo) [reply src/client/datascience/data-viewing/jupyterVariableDataProvider.ts:57]: Yeah ```if (column.describe)``` is undefined, so ```col['describe']``` never gets created. Will that break something else later?

Comment (vandyliu) [reply src/client/datascience/data-viewing/jupyterVariableDataProvider.ts:57]: I think it's okay because it is potentially undefined from the type definition. I can play it safe and give it an empty string though.

File: src/client/datascience/interactive-common/interactiveBase.ts (hunk 24 above)
Comment (vandyliu) [reply src/client/datascience/interactive-common/interactiveBase.ts:340]: WebviewPanelHost or WebViewHost would be better?
WebviewPanelHost extends WebViewHost
WebViewHost also has 4 other onMessages already: SharedMessages.Started, CssMessages.GetCssRequest, CssMessages.GetMonacoThemeRequest, and InteractiveWindowMessages.GetHTMLByIdResponse

Leaning towards WebViewHost but I don't have enough knowledge about it

Comment (rchiodo) [reply src/client/datascience/interactive-common/interactiveBase.ts:340]: I think the lower you can push this stuff the better.

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 25]
```diff
@@ -0,0 +1,821 @@
...
+// Runs the transformations, communicates with the data wrangler UI through onMessage and postMessage
+@injectable()
+export class DataWrangler extends DataViewer implements IDataWrangler, IDisposable {
+    private existingDisposable: Disposable | undefined;
+    private historyList: IHistoryItem[] = [];
+    private sourceFile: string | undefined;
+    private commands = new Map<
+        DataWranglerCommands,
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (args: any, currentVariableName: string) => Promise<IHistoryItem | void>
```
Comment (pwang347) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:69]: nit: if you want typed arguments you could possibly do something like

```
export type DataWranglerCommandArgs<T extends DataWranglerCommands> = 
  T extends DataWranglerCommands.Describe ?
  IDescribeArgs :
  T extends DataWranglerCommands.Drop ?
  IDropArgs :
  any;

private commands = new Map<
        DataWranglerCommands,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        <T>(args: DataWranglerCommandArgs<T>, currentVariableName: string) => Promise<IHistoryItem | void>
```

another option would be to add the command to the args interface as well, and have a type that is like
```
export type DataWranglerCommandArgs = IDescribeArgs | IDropArgs;
```

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 26]
```diff
@@ -0,0 +1,821 @@
...
+                    DataWranglerMessages.SetSidePanels,
+                    wantedPanels as SidePanelSections[]
+                ).ignoreErrors();
+            }
+
+            // Use Data Viewer logic to show initial data
+            const dataFrameInfo = await this.showInitialData(title);
+            this.sourceFile = dataFrameInfo.sourceFile;
+
+            this.historyList.push({
```
Comment (pwang347) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:162]: nit: should we use the `addToHistory` helper here?

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 27]
```diff
@@ -0,0 +1,821 @@
...
+                description: DataScience.dataWranglerImportDescription(),
+                code: `import pandas as pd\r\ndf = pd.read_csv(r'${this.sourceFile ?? ''}')\n`,
```
Comment (pwang347) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:164]: nit: can we have a helper for multiline codegen e.g. something like
```
createCode([
  "import pandas as pd",
  `df = pd.read_csv(r'${this.sourceFile ?? '''}')`
], { lineEnding: <override default?> })
```

Comment (pwang347) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:164]: I think we should probably default to the system line ending, ie from `TextDocument.eol`

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 28]
```diff
@@ -0,0 +1,821 @@
...
+        this.commands.set(DataWranglerCommands.CoerceColumn, this.coerceColumn.bind(this));
+        this.commands.set(DataWranglerCommands.ReplaceAllColumn, this.replaceAllColumn.bind(this));
+        this.commands.set(DataWranglerCommands.RemoveHistoryItem, this.removeHistoryItem.bind(this));
+        this.commands.set(DataWranglerCommands.ExportToCsv, this.exportToCsv.bind(this));
+        this.commands.set(DataWranglerCommands.RespondToPreview, this.respondToPreview.bind(this));
+
+        this.onDidDispose(this.dataWranglerDisposed, this);
+    }
+
+    public async showData(
```
Comment (pwang347) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:138]: can this be merged with `showInitialData`?

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 29]
```diff
@@ -0,0 +1,821 @@
...
+                handled = true;
+                break;
+
+            case DataWranglerMessages.RefreshDataWrangler:
+                this.refreshData().ignoreErrors();
+                handled = true;
+                break;
+
+            case CssMessages.GetMonacoThemeRequest:
+                void this.handleMonacoThemeRequest(payload);
```
Comment (pwang347) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:241]: is the void here a typo?

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 30]
```diff
@@ -0,0 +1,821 @@
...
+        }
+    }
+
+    private addToHistory(newHistoryItem: IHistoryItem) {
+        this.historyList.push(newHistoryItem);
+        this.postMessage(DataWranglerMessages.UpdateHistoryList, this.historyList).ignoreErrors();
+    }
+
+    private getCode() {
+        return this.historyList.map((item) => item.code).join('\n');
```
Comment (pwang347) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:261]: same nit, maybe use the document EOL

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 31]
```diff
@@ -0,0 +1,821 @@
...
+
+    private async exportToCsv(_req: undefined, currentVariableName: string) {
+        const notebook = (this.dataProvider as IJupyterVariableDataProvider).notebook;
+        const fileInfo = await this.applicationShell.showSaveDialog({
+            saveLabel: DataScience.dataWranglerSaveCsv(),
+            filters: { CSV: ['csv'] }
+        });
+        if (fileInfo) {
+            const code = `${currentVariableName}.to_csv(path_or_buf=r'${fileInfo.fsPath}', index=False)`;
+            await notebook?.execute(code, '', 0, uuid(), undefined, false);
```
Comment (pwang347) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:272]: I would recommend writing a helper over this function, since you really only care about the `code` part as a parameter

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 32]
```diff
@@ -0,0 +1,821 @@
...
+    private async handleCommand(payload: { command: DataWranglerCommands; args: any }) {
+        const notebook = (this.dataProvider as IJupyterVariableDataProvider).notebook;
+        let codeToRun;
+        const currentVariableName = (await this.dataFrameInfoPromise)!.name ?? '';
+        let newVariableName = currentVariableName ?? '';
+        let historyItem: IHistoryItem | void;
+
+        // Get and run data wrangler command
+        const cmd = this.commands.get(payload.command);
+        if (cmd) {
```
Comment (pwang347) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:305]: nit: maybe do
```
if (!cmd) {
  return;
}

...
```

so we know there's nothing left to check for that case below

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 33]
```diff
@@ -0,0 +1,821 @@
...
+            historyItem = await cmd(payload.args, currentVariableName);
+            if (historyItem !== undefined) {
+                codeToRun = historyItem.isPreview ? historyItem.previewCode : historyItem.code;
```
Comment (pwang347) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:308]: nit: might be good to add a small comment here

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 34]
```diff
@@ -0,0 +1,821 @@
...
+        const dataCleanCode = this.getCode();
+        const notebookEditor = await this.commandManager.executeCommand(Commands.CreateNewNotebook);
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const blankCell = (notebookEditor as any).document.cellAt(0) as NotebookCell;
+        await updateCellCode(blankCell, dataCleanCode);
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private async handleCommand(payload: { command: DataWranglerCommands; args: any }) {
+        const notebook = (this.dataProvider as IJupyterVariableDataProvider).notebook;
```
Comment (pwang347) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:297]: curious, can we just store this value once and make sure it's not undefined? do we need to keep getting references and checking for undefined?

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts [hunk 35]
```diff
@@ -0,0 +1,821 @@
...
+            if (this.existingDisposable) {
+                this.existingDisposable.dispose();
+            }
+            if (newVariableName) {
+                await this.updateWithNewVariable(newVariableName);
+            }
+        }
+
+        if (historyItem) {
+            if (historyItem?.shouldAdd) {
```
Comment (pwang347) [inline src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:325]: nit: we don't need these `?`s after the check above

**Review by pwang347 — COMMENTED** [review]
will continue review in a bit

File: src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts (hunk 34 above)
Comment (vandyliu) [reply src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts:297]: I believe it changes when `updateWithNewVariable` is called
